lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -25,83 +25,105 @@ def __init__(self, context):
-                self.logger.info(
-                    'All tools removed from group "%s" on host "%s"',
-                    self.context.group,
-                    tg_dir_r.name,
-                )
+                    continue
+
+                if self.context.name:
+                    names = self.context.name.split(",")
+                else:
+                    # Discover all the tools registered for this remote
+                    names = self.tools(tg_dir_r)
+                    if not names:
+                        # FIXME:  this is another odd case -- the remote subdirectory
+                        #  exists, but it's empty.  (We'll remove it below.)
+                        self.logger.warn(
+                            'No tools in group "%s" on host "%s".', group, remote,
+                        )
+
+                for name in names:
+                    status = self._clear_tools(name, remote, group)
+                    if status > 0:
+                        errors = 1
+                        tools_not_found.append(name)
+                    elif status < 0:
+                        tools_not_found.append(name)
+
+                tool_files = [p.name for p in tg_dir_r.iterdir()]
+
+                if len(tool_files) == 1 and tool_files[0] == "__label__":
+                    label = tg_dir_r / "__label__"
+                    try:
+                        label.unlink()
+                    except Exception:
+                        self.logger.error(
+                            "Failed to remove label for remote %s", tg_dir_r
+                        )
+                        errors = 1
+
+                if self.is_empty(tg_dir_r):
+                    self.logger.info(
+                        'All tools removed from group "%s" on host "%s"',
+                        group,
+                        tg_dir_r.name,
+                    )
+                    try:
+                        shutil.rmtree(tg_dir_r)
+                    except OSError:
+                        self.logger.error(
+                            "Failed to remove remote directory %s", tg_dir_r
+                        )
+                        errors = 1
+
+            if self.context.name and tools_not_found:
+                self.logger.warn(f"Tools {tools_not_found} not found")
+
+            if group != "default" and not any(self.tool_group_dir.iterdir()):
                 try:
-                    shutil.rmtree(tg_dir_r)
+                    shutil.rmtree(self.tool_group_dir)
                 except OSError:
-                    self.logger.error("Failed to remove remote directory %s", tg_dir_r)
+                    self.logger.error(
+                        "Failed to remove group directory %s", self.tool_group_dir
+                    )
                     errors = 1
 
-        if self.context.name and tool_not_found:
-            self.logger.warn(f'Tool "{self.context.name}" not found')
-
         return 0 if errors == 0 else 1
```

Comment:
Since `errors` isn't a *counter* but just an integer set to 0 or 1, this might as well be `return errors`.

Reply:
I own this one.  Prior to my last change, `errors` was in fact a counter (although, the only values which mattered were zero and "not-zero", so  I changed it to a "flag" (I am so not enamored of that construct!)...but I didn't go so far as to actually make it a boolean (which I probably should have)).  I then toyed with the idea of returning the value of `errors` directly, but then the moral part of me got upset because that would be returning what was in essence a boolean value through an interface contract which called for an `int`.   So, rather than changing the interface or putting in a typecast (or relying on an implicit one), I used that silly-looking ternary.  🦐 

If you want to just return the value of `errors` directly, I won't object.